kimchi/src/alphas.rs

```diff
@@ -0,0 +1,276 @@
+//! This module implements an abstraction to keep track of the powers of alphas.
+//! As a recap, alpha is a challenge sent by the verifier in PLONK,
+//! and is used to segregate different constraints.
+//! It is important that different constraints use different powers of alpha,
+//! as otherwise they can interact and potentially cancel one another.
+//! (The proof is in the use of the Schwartz-Zipple lemma.)
+//! As such, we want two properties from this:
+//!
+//! - we should keep track of a mapping between type of constraint and range of powers
+//! - when powers of alphas are used, we should ensure that no more no less are used
+//!
+//! We use powers of alpha in two different places in the codebase:
+//!
+//! - when creating the index, we do not know alpha at this point so we
+//!   simply keep track of what constraints will use what powers
+//! - when creating a proof or verifying a proof, at this point we know alpha
+//!   so we can use the mapping we created during the creation of the index.
+//!
+//! For this to work, we use two types:
+//!
+//! - [Builder], which allows us to map constraints to powers
+//! - [Alphas], which you can derive from [Builder] and an `alpha`
+//!
+//! Both construction will enforce that you use all the powers of
+//! alphas that you register for constraint. This allows us to
+//! make sure that we compute the correct amounts, without reusing
+//! powers of alphas between constraints.
+
+use ark_ff::Field;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    iter::{Cloned, Take},
+    ops::Range,
+    slice::Iter,
+    thread,
+};
+
+// ------------------------------------------
+
+/// A constraint type represents a polynomial that will be part of the final equation f (the circuit equation)
+#[derive(PartialEq, Eq, Clone, Hash, Debug, Serialize, Deserialize)]
+pub enum ConstraintType {
+    /// gates in the PLONK constraint system.
+    /// As gates are mutually exclusive (only a single selector polynomial
+    /// will be non-zero for a given point),
+    /// we can reuse the same power of alphas accross gates.
+    Gate,
+    /// The permutation argument
+    Permutation,
+    /// The lookup argument
+    Lookup,
+}
+
+// ------------------------------------------
+
+/// This type can be used to create a mapping between powers of alpha and constraint types.
+/// See [Builder::default] to create one,
+/// and [Builder::register] to register a new mapping.
+/// Once you know the alpha value, you can convert this type to a [Alphas].
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+pub struct Builder {
+    /// The next power of alpha to use
+    next_power: usize,
```

Comment:
I thought the comment and the "next" were enough to clarify that :o otherwise I would have called that "last"

Reply:
mmm well perhaps this is just me? the name felt a bit confusing to me. So if you say "last" to me it will sound like "the very last one that is included".